src/goto-diff/goto_diff_base.cpp

```diff
@@ -11,95 +11,134 @@ Author: Peter Schrammel
 
 #include "goto_diff.h"
 
+#include <goto-programs/show_properties.h>
+
 #include <util/json_expr.h>
+#include <util/options.h>
+
 
-std::ostream &goto_difft::output_functions(std::ostream &out) const
+/// Output diff result
+void goto_difft::output_functions() const
 {
-  namespacet ns1(goto_model1.symbol_table);
-  namespacet ns2(goto_model2.symbol_table);
   switch(ui)
   {
     case ui_message_handlert::uit::PLAIN:
     {
-      out << "total number of functions: " << total_functions_count << "\n";
-      out << "new functions:\n";
-      for(irep_id_sett::const_iterator it=new_functions.begin();
-          it!=new_functions.end(); ++it)
-      {
-        const symbolt &symbol = ns2.lookup(*it);
-        out << "  " << symbol.location.get_file() << ": " << *it << "\n";
-      }
-
-      out << "modified functions:\n";
-      for(irep_id_sett::const_iterator it=modified_functions.begin();
-          it!=modified_functions.end(); ++it)
-      {
-        const symbolt &symbol = ns2.lookup(*it);
-        out << "  " << symbol.location.get_file() << ": " << *it << "\n";
-      }
-
-      out << "deleted functions:\n";
-      for(irep_id_sett::const_iterator it=deleted_functions.begin();
-          it!=deleted_functions.end(); ++it)
-      {
-        const symbolt &symbol = ns1.lookup(*it);
-        out << "  " << symbol.location.get_file() << ": " << *it << "\n";
-      }
+      result() << "total number of functions: " << total_functions_count << '\n';
+      output_function_group("new functions", new_functions, goto_model2);
+      output_function_group("modified functions", modified_functions, goto_model2);
+      output_function_group("deleted functions", deleted_functions, goto_model1);
+      result() << eom;
       break;
     }
     case ui_message_handlert::uit::JSON_UI:
     {
       json_objectt json_result;
       json_result["totalNumberOfFunctions"]=
         json_stringt(std::to_string(total_functions_count));
-      convert_function_group
-        (json_result["newFunctions"].make_array(), new_functions);
-      convert_function_group(
-        json_result["modifiedFunctions"].make_array(), modified_functions);
-      convert_function_group(
-        json_result["deletedFunctions"].make_array(), deleted_functions);
-      out << ",\n" << json_result;
+      convert_function_group_json(
+        json_result["newFunctions"].make_array(),
+        new_functions, goto_model2);
+      convert_function_group_json(
+        json_result["modifiedFunctions"].make_array(),
+        modified_functions, goto_model2);
+      convert_function_group_json(
+        json_result["deletedFunctions"].make_array(),
+        deleted_functions, goto_model1);
+      result() << json_result;
       break;
     }
     case ui_message_handlert::uit::XML_UI:
     {
-      out << "not supported yet";
+      error() << "not supported yet" << eom;
```

Comment:
Would you mind changing the error message as well to "XML output not yet supported"?